homeassistant/helpers/singleton.py

```diff
@@ -25,6 +25,7 @@ def wrapper(func: _FuncType[_T]) -> _FuncType[_T]:
         """Wrap a function with caching logic."""
         if not asyncio.iscoroutinefunction(func):
 
+            @functools.lru_cache(maxsize=1)
```

Comment:
By wrapping in an lru_cache at the end we preserve compat since the data still ends up in `hass.data`, but when this is called it will not have to run python code as it will hit the native code implementation.